hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/inject/AvroUpsertTarget.java

```diff
@@ -16,20 +16,45 @@
 
 package com.hazelcast.jet.sql.impl.inject;
 
+import com.hazelcast.internal.util.collection.DefaultedMap;
 import com.hazelcast.sql.impl.QueryException;
 import com.hazelcast.sql.impl.type.QueryDataType;
 import org.apache.avro.Schema;
-import org.apache.avro.generic.GenericContainer;
 import org.apache.avro.generic.GenericData.Record;
 import org.apache.avro.generic.GenericRecordBuilder;
 
 import javax.annotation.Nullable;
 import javax.annotation.concurrent.NotThreadSafe;
 
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.Map;
+
+import static com.hazelcast.jet.sql.impl.connector.file.AvroResolver.AVRO_TO_SQL;
+import static com.hazelcast.jet.sql.impl.connector.file.AvroResolver.unwrapNullableType;
 import static com.hazelcast.jet.sql.impl.inject.UpsertInjector.FAILING_TOP_LEVEL_INJECTOR;
+import static org.apache.avro.Schema.Type.BOOLEAN;
+import static org.apache.avro.Schema.Type.DOUBLE;
+import static org.apache.avro.Schema.Type.FLOAT;
+import static org.apache.avro.Schema.Type.INT;
+import static org.apache.avro.Schema.Type.LONG;
+import static org.apache.avro.Schema.Type.STRING;
 
 @NotThreadSafe
 class AvroUpsertTarget implements UpsertTarget {
+    // In integers and decimals, we try to preserve the precision.
+    // In floating-point numbers, we try to preserve the scale since they can
+    // underflow, i.e. become 0 due to being very small, when converted to an integer.
+    private static final Map<Class<?>, List<Schema.Type>> CONVERSION_PREFS = new DefaultedMap<>(Map.of(
+            Boolean.class, List.of(BOOLEAN, STRING),
```

Comment:
are you sure we should fallback to STRING for all types? Do we do that in other cases? String for numbers (also time) is tricky due to locale-specific format.

Reply:
> `KvMetadataAvroResolver` and `AvroUpsertTarget` support all conversions that are defined by [converters][1].
1. Converters parse and produce ISO-formatted date/time, so locale is ignored.
2. Converters support every possible conversion from/to `VARCHAR`.

Here, my intention is not to put any artificial limits on what we can do with Avro. Other serialization formats are more strict about conversions, but this is because we have control over them. For example, `BigDecimal` can be represented in Java/Compact/Portable as-is, whereas in Avro, it should either be converted into `String` or [`byte[]`][2].

[1]: https://github.com/hazelcast/hazelcast/blob/972078db18e2ed7bb9df88021b33382b3231ab04/hazelcast-sql/src/main/java/com/hazelcast/sql/impl/type/converter
[2]: https://avro.apache.org/docs/1.11.1/specification/#decimal

Reply:
thanks for explanation